src/Microsoft.AspNet.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -0,0 +1,550 @@
+
+using System;
+using System.Globalization;
+using System.Text;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNet.Mvc.Formatters
+{
+    /// <summary>
+    /// A media type value.
+    /// </summary>
+    public struct MediaType
+    {
+        private static readonly StringSegment QualityParameter = new StringSegment("q");
+
+        private MediaTypeParameterParser _parameterParser;
+
+        /// <summary>
+        /// Initializes a <see cref="MediaType"/> instance.
+        /// </summary>
+        /// <param name="mediaType">The <see cref="string"/> with the media type.</param>
+        public MediaType(string mediaType)
+            : this(mediaType, 0, mediaType.Length)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a <see cref="MediaType"/> instance.
+        /// </summary>
+        /// <param name="mediaType">The <see cref="StringSegment"/> with the media type.</param>
+        public MediaType(StringSegment mediaType)
+            : this(mediaType.Buffer, mediaType.Offset, mediaType.Length)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a <see cref="MediaTypeParameterParser"/> instance.
+        /// </summary>
+        /// <param name="mediaType">The <see cref="string"/> with the media type.</param>
+        /// <param name="offset">The offset in the <paramref name="mediaType"/> where the parsing starts.</param>
+        /// <param name="length">The of the media type to parse if provided.</param>
+        public MediaType(string mediaType, int offset, int? length)
+        {
+            if (mediaType == null)
+            {
+                throw new ArgumentNullException(nameof(mediaType));
+            }
+
+            if (offset < 0 || offset >= mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (length != null && offset + length > mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            _parameterParser = default(MediaTypeParameterParser);
+            Type = new StringSegment();
+            SubType = new StringSegment();
+
+            StringSegment type;
+            var typeLength = GetTypeLength(mediaType, offset, out type);
+            if (typeLength == 0)
+            {
+                return;
+            }
+            else
+            {
+                Type = type;
+            }
+
+            StringSegment subType;
+            var subtypeLength = GetSubtypeLength(mediaType, offset + typeLength, out subType);
+            if (subtypeLength == 0)
+            {
+                return;
+            }
+            else
+            {
+                SubType = subType;
+            }
+
+            _parameterParser = new MediaTypeParameterParser(mediaType, offset + typeLength + subtypeLength, length);
```

Comment:
offset + typeLength + subtypeLenght is correct, GetXXXLength method will return how many characters the parser has advanced, including tokens like '/' and trivia like whitespaces, this accounts for situations like application / json which we parse successfully (as we did before this change)